instana/resource-sli-config.go

```diff
@@ -0,0 +1,261 @@
+				SliConfigFieldEndpointID: {
+					Type:        schema.TypeString,
+					Optional:    true,
+					Description: "The endpoint ID of the entity",
+				},
+				//SliConfigFieldBoundaryScope nested schema field definition of instana_sli_config field sli_entity.boundary_scope
+				SliConfigFieldBoundaryScope: {
+					Type:         schema.TypeString,
+					Required:     true,
+					ValidateFunc: validation.StringInSlice([]string{"ALL", "INBOUND"}, true),
+					Description:  "The boundary scope for the entity configuration (ALL, INBOUND)",
+				},
+			},
+		},
+	}
+)
+
+//NewSliConfigResourceHandle creates the resource handle for SLI configuration
+func NewSliConfigResourceHandle() *ResourceHandle {
+	return &ResourceHandle{
+		ResourceName: ResourceInstanaSliConfig,
+		Schema: map[string]*schema.Schema{
+			SliConfigFieldName:                       SliConfigName,
+			SliConfigFieldFullName:                   SliConfigFullName,
+			SliConfigFieldInitialEvaluationTimestamp: SliConfigInitialEvaluationTimestamp,
+			SliConfigFieldMetricConfiguration:        SliConfigMetricConfiguration,
+			SliConfigFieldSliEntity:                  SliConfigSliEntity,
+		},
+		SchemaVersion: 1,
+		StateUpgraders: []schema.StateUpgrader{
+			{
+				Type:    sliConfigSchemaV0().CoreConfigSchema().ImpliedType(),
+				Upgrade: sliConfigStateUpgradeV0,
+				Version: 0,
+			},
+		},
+		RestResourceFactory:  func(api restapi.InstanaAPI) restapi.RestResource { return api.SliConfigs() },
+		UpdateState:          updateStateForSliConfig,
+		MapStateToDataObject: mapStateToDataObjectForSliConfig,
+	}
+}
+
+func updateStateForSliConfig(d *schema.ResourceData, obj restapi.InstanaDataObject) error {
+	sliConfig := obj.(restapi.SliConfig)
+
+	metricConfiguration := map[string]string{
+		SliConfigFieldMetricName:        sliConfig.MetricConfiguration.Name,
+		SliConfigFieldMetricAggregation: sliConfig.MetricConfiguration.Aggregation,
+		SliConfigFieldMetricThreshold:   fmt.Sprintf("%f", sliConfig.MetricConfiguration.Threshold),
+	}
+
+	sliEntity := map[string]string{
+		SliConfigFieldSliType:       sliConfig.SliEntity.Type,
+		SliConfigFieldApplicationID: sliConfig.SliEntity.ApplicationID,
+		SliConfigFieldServiceID:     sliConfig.SliEntity.ServiceID,
+		SliConfigFieldEndpointID:    sliConfig.SliEntity.EndpointID,
+		SliConfigFieldBoundaryScope: sliConfig.SliEntity.BoundaryScope,
+	}
+
+	d.Set(SliConfigFieldFullName, sliConfig.Name)
+	d.Set(SliConfigFieldInitialEvaluationTimestamp, sliConfig.InitialEvaluationTimestamp)
+	d.Set(SliConfigFieldMetricConfiguration, metricConfiguration)
+	d.Set(SliConfigFieldSliEntity, sliEntity)
+
+	d.SetId(sliConfig.ID)
+	return nil
+}
+
+func mapStateToDataObjectForSliConfig(d *schema.ResourceData, formatter utils.ResourceNameFormatter) (restapi.InstanaDataObject, error) {
+	metricConfigurationsList := d.Get(SliConfigFieldMetricConfiguration).([]interface{})
+	var metricConfiguration restapi.MetricConfiguration
+	if len(metricConfigurationsList) > 0 {
```

Comment:
could be moved into a separate method mapMetricConfigurationEntityFromState()

Reply:
refactored into separate method